celery/app/amqp.py

```diff
@@ -521,7 +521,7 @@ def send_task_message(producer, name, message,
                     exchange_type = 'direct'
 
             # convert to anon-exchange, when exchange not set and direct ex.
-            if not exchange or not routing_key and exchange_type == 'direct':
+            if (not exchange or not routing_key) and exchange_type == 'direct':
```

Comment:
I'll give it a try, any pointers of where to add this / look for a similar test?